stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,326 @@
+//===--- StaticBinaryELF.cpp ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Parse a static ELF binary to implement lookupSymbol() address lookup.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__)
+
+#include "ImageInspection.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Object/ELF.h"
+#include "swift/Basic/Lazy.h"
+#include <cassert>
+#include <string>
+#include <vector>
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <linux/limits.h>
+#include <sys/auxv.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+using namespace std;
+using namespace llvm;
+using elfType = object::ELFType<support::native, (__POINTER_WIDTH__ == 64)>;
+
+extern const elfType::Ehdr elfHeader asm("__ehdr_start");
+
+// Create strong linkage to pthread_self, pthread_once and pthread_key_create
+// as they are normally weak-linked and used to detect the presence of pthreads.
+// Without this the calls just jump to 0x0.
+__attribute__((__visibility__("hidden")))
+ pthread_t (*__strong_pthread_self)(void) = pthread_self;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_once)(pthread_once_t *, void (*)(void)) = pthread_once;
+
+__attribute__((__visibility__("hidden")))
+int (*__strong_pthread_key_create)(pthread_key_t *, void (*)(void *)) = pthread_key_create;
```

Comment:
I really wish we had linker options ... `-u pthread_self -u pthread_once -u pthead_key_create` is so much better than this.  *sigh*

Reply:
I had used `---defsym=__import_pthread_self=pthread_self` options to the linker before but it looks like that doesnt create strong linkage anymore.

Reply:
`--defsym` creating a strong reference sounds like a bug.  If you want to create a strong reference, you should use `--require-defined` or `-u` at the very least.

Reply:
I didnt know about this option! I had to use `-undefined (-u)` since `ld.gold` is being used, `--require-defined` seems to only work on `ld`